File: src/manager/task_manager_python/task_manager_python/task_manager_node.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import rclpy
from rclpy.node import Node
from std_msgs.msg import String, Bool, Empty, Float32
from collections import deque

class TaskManager(Node):
    def __init__(self):
        super().__init__('task_manager')

        # 状態の定義
        self.STATE_INITIALIZING = "initializing"
        self.STATE_SEARCHING = "searching"
        self.STATE_APPROACHING = "approaching"
        self.STATE_COLLECTING = "collecting"
        self.STATE_STOPPING = "stopping"
        
        self._state = ""
        self.target_queue = deque()
        
        # 状態遷移のためのタイマーを保持する変数
        self.transition_timer = None

        # Publisher / Subscriber
        self.state_pub = self.create_publisher(String, '/robot/state', 10)
        self.camera_angle_pub = self.create_publisher(Float32, '/cameraswingmotor/target_angle', 10)
        self.create_subscription(Empty, '/detected_objects_3d', self.detected_objects_callback, 10)
        self.create_subscription(Bool, '/hose/result', self.hose_result_callback, 10)

        # 状態を定期的に配信するタイマー
        self.create_timer(1.0, self.publish_state)

        self.get_logger().info("✅ Task Manager ノード (ROS2) が起動しました。")
        
        # 1. 初期化状態から開始
        self.set_state(self.STATE_INITIALIZING)

        # 初期化時にカメラの角度を設定
        self.initialize_camera()
        
        # 2. 2秒後に探索を開始
        self.transition_timer = self.create_timer(2.0, self.start_searching)

    def initialize_camera(self):
        self.get_logger().info("カメラの初期化を行います...")
        self.initial_angle_timer = self.create_timer(0.5, self._publish_initial_camera_angle)

    def _publish_initial_camera_angle(self):
        initial_angle = Float32()
        initial_angle.data = 20.0  # 初期角度を20度に設定
        self.camera_angle_pub.publish(initial_angle)
        self.get_logger().info(f"カメラの角度を {initial_angle.data} 度に設定しました。")

        if self.initial_angle_timer is not None:
            self.initial_angle_timer.cancel()

    def set_state(self, new_state):
        if self._state != new_state:
            self._state = new_state
            self.get_logger().info(f"====== 状態が [ {self._state.upper()} ] に遷移しました ======")
            self.publish_state()

            # 進行中の遷移タイマーがあればキャンセル
            if self.transition_timer is not None and not self.transition_timer.is_canceled():
                self.transition_timer.cancel()

            # 状態に応じた処理を開始
            if self._state == self.STATE_APPROACHING:
                self.get_logger().info("  -> ゴミに接近中...")
                # TODO: ナビゲーションノードへ目標を指示
                self.transition_timer = self.create_timer(3.0, self.on_approach_done)

            elif self._state == self.STATE_COLLECTING:
                self.get_logger().info("  -> ホースでゴミを回収中...")
                # TODO: ホース制御ノードへ回収開始を指示

            elif self._state == self.STATE_STOPPING:
                self.get_logger().info("  -> 一時停止中...")
                self.transition_timer = self.create_timer(1.0, self.on_stop_done)
                
    def publish_state(self):
        msg = String()
        msg.data = self._state
        self.state_pub.publish(msg)

    def start_searching(self):
        """初期化完了後、探索状態に移行するコールバック"""
        self.transition_timer.cancel()
        self.set_state(self.STATE_SEARCHING)

    def on_approach_done(self):
        """接近完了（シミュレーション）時に呼ばれるコールバック"""
        self.transition_timer.cancel()
        if self._state == self.STATE_APPROACHING:
            self.set_state(self.STATE_COLLECTING)

    def on_stop_done(self):
        """停止処理完了時に呼ばれるコールバック"""
        self.transition_timer.cancel()
        if self._state == self.STATE_STOPPING:
            if self.target_queue:
                self.target_queue.popleft()
            
            if self.target_queue:
                self.get_logger().info("次のゴミへ移動します。")
                self.set_state(self.STATE_APPROACHING)
            else:
                self.get_logger().info("全てのゴミを回収しました。探索を再開します。")
                self.set_state(self.STATE_SEARCHING)

    def detected_objects_callback(self, msg):
        if self._state == self.STATE_SEARCHING:
            self.get_logger().info("🗑️ ゴミを検出しました。タスクキューに追加します。")
            self.target_queue.append("dummy_target_1")
            self.set_state(self.STATE_APPROACHING)

    def hose_result_callback(self, msg):
        if self._state == self.STATE_COLLECTING:
            if msg.data:
                self.get_logger().info("👍 回収成功！")
            else:
                self.get_logger().info("😥 回収失敗...")
            self.set_state(self.STATE_STOPPING)

def main(args=None):
    rclpy.init(args=args)
    task_manager_node = TaskManager()
    try:
        rclpy.spin(task_manager_node)
    except KeyboardInterrupt:
        pass
    finally:
        task_manager_node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/manager/task_manager/src/task_manager_node.cpp
#include <chrono>
#include <memory>
#include <string>
#include <deque>
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <std_msgs/msg/bool.hpp>
#include <std_msgs/msg/float32.hpp>
#include <geometry_msgs/msg/point_stamped.hpp>

using namespace std::chrono_literals;

class TaskManagerNode : public rclcpp::Node {
public:
  TaskManagerNode()
  : rclcpp::Node("task_manager")
  {
    RCLCPP_INFO(this->get_logger(), "✅ Task Manager ノード (C++) を起動しました");

    // パラメータ宣言
    this->declare_parameter<double>("initial_angle_deg", 20.0);
    this->declare_parameter<double>("initial_angle_delay_sec", 0.5);
    this->declare_parameter<double>("search_start_delay_sec", 2.0);
    this->declare_parameter<double>("state_publish_period_sec", 1.0);
    this->declare_parameter<double>("stop_duration_sec", 1.0);

    initial_angle_deg_ = this->get_parameter("initial_angle_deg").as_double();
    initial_angle_delay_sec_ = this->get_parameter("initial_angle_delay_sec").as_double();
    search_start_delay_sec_ = this->get_parameter("search_start_delay_sec").as_double();
    state_publish_period_sec_ = this->get_parameter("state_publish_period_sec").as_double();
    stop_duration_sec_ = this->get_parameter("stop_duration_sec").as_double();

    // 状態定義
    STATE_INITIALIZING = "initializing";
    STATE_SEARCHING = "searching";
    STATE_APPROACHING = "approaching";
    STATE_COLLECTING = "collecting";
    STATE_STOPPING = "stopping";

    // Publisher / Subscriber
    state_pub_ = this->create_publisher<std_msgs::msg::String>("/robot/state", 10);
    camera_angle_pub_ = this->create_publisher<std_msgs::msg::Float32>("/cameraswingmotor/target_angle", 10);

    detected_sub_ = this->create_subscription<geometry_msgs::msg::PointStamped>(
      "/detected_depth_points", 10,
      std::bind(&TaskManagerNode::detected_objects_callback, this, std::placeholders::_1));

    hose_result_sub_ = this->create_subscription<std_msgs::msg::Bool>(
      "/hose/result", 10,
      std::bind(&TaskManagerNode::hose_result_callback, this, std::placeholders::_1));

    chaser_completion_sub_ = this->create_subscription<std_msgs::msg::Bool>(
      "/chaser/approach_completed", 10,
      std::bind(&TaskManagerNode::chaser_completion_callback, this, std::placeholders::_1));

    // 状態を定期的に配信するタイマー
    state_publish_timer_ = this->create_wall_timer(
      std::chrono::duration<double>(state_publish_period_sec_),
      std::bind(&TaskManagerNode::publish_state, this));

    // 初期状態
    set_state(STATE_INITIALIZING);

    // カメラ初期化（不要なのでコメントアウト）
    // initialize_camera();

    // 2秒後に探索開始
    transition_timer_ = this->create_wall_timer(
      std::chrono::duration<double>(search_start_delay_sec_),
      std::bind(&TaskManagerNode::start_searching, this));
  }

private:
  // 状態
  std::string STATE_INITIALIZING;
  std::string STATE_SEARCHING;
  std::string STATE_APPROACHING;
  std::string STATE_COLLECTING;
  std::string STATE_STOPPING;

  std::string state_;
  std::deque<std::string> target_queue_;

  // パラメータ値
  double initial_angle_deg_{};
  double initial_angle_delay_sec_{};
  double search_start_delay_sec_{};
  double state_publish_period_sec_{};
  double stop_duration_sec_{};

  // Publishers / Subscribers
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr state_pub_;
  rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr camera_angle_pub_;
  rclcpp::Subscription<geometry_msgs::msg::PointStamped>::SharedPtr detected_sub_;
  rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr hose_result_sub_;
  rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr chaser_completion_sub_;

  // Timers
  rclcpp::TimerBase::SharedPtr state_publish_timer_;
  rclcpp::TimerBase::SharedPtr transition_timer_;
  rclcpp::TimerBase::SharedPtr initial_angle_timer_;

  void initialize_camera()
  {
    RCLCPP_INFO(this->get_logger(), "カメラの初期化を行います...");
    initial_angle_timer_ = this->create_wall_timer(
      std::chrono::duration<double>(initial_angle_delay_sec_),
      std::bind(&TaskManagerNode::_publish_initial_camera_angle, this));
  }

  void _publish_initial_camera_angle()
  {
    std_msgs::msg::Float32 msg;
    msg.data = static_cast<float>(initial_angle_deg_);
    camera_angle_pub_->publish(msg);
    RCLCPP_INFO(this->get_logger(), "カメラの角度を %f 度に設定しました。", msg.data);
    if (initial_angle_timer_ && !initial_angle_timer_->is_canceled()) {
      initial_angle_timer_->cancel();
    }
  }

  void set_state(const std::string & new_state)
  {
    if (state_ == new_state) return;
    state_ = new_state;
    RCLCPP_INFO(this->get_logger(), "====== 状態が [ %s ] に遷移しました ======", state_.c_str());
    publish_state();

    if (transition_timer_ && !transition_timer_->is_canceled()) {
      transition_timer_->cancel();
    }

    if (state_ == STATE_APPROACHING) {
      RCLCPP_INFO(this->get_logger(), "  -> ゴミに接近中...ObjectChaserNodeからの完了通知を待ちます。");
      // 必要ならここでナビゲーションへ目標を指示する。
    } else if (state_ == STATE_COLLECTING) {
      RCLCPP_INFO(this->get_logger(), "  -> ホースでゴミを回収中...");
      // ホース制御へ回収開始を指示する。
    } else if (state_ == STATE_STOPPING) {
      RCLCPP_INFO(this->get_logger(), "  -> 一時停止中...");
      transition_timer_ = this->create_wall_timer(
        std::chrono::duration<double>(stop_duration_sec_),
        std::bind(&TaskManagerNode::on_stop_done, this));
    }
  }

  void publish_state()
  {
    std_msgs::msg::String msg;
    msg.data = state_;
    state_pub_->publish(msg);
  }

  void start_searching()
  {
    if (transition_timer_ && !transition_timer_->is_canceled()) {
      transition_timer_->cancel();
    }
    set_state(STATE_SEARCHING);
  }

  void on_stop_done()
  {
    if (transition_timer_ && !transition_timer_->is_canceled()) {
      transition_timer_->cancel();
    }
    if (state_ == STATE_STOPPING) {
      if (!target_queue_.empty()) {
        target_queue_.pop_front();
      }
      if (!target_queue_.empty()) {
        RCLCPP_INFO(this->get_logger(), "次のゴミへ移動します。");
        set_state(STATE_APPROACHING);
      } else {
        RCLCPP_INFO(this->get_logger(), "全てのゴミを回収しました。探索を再開します。");
        set_state(STATE_SEARCHING);
      }
    }
  }

  void detected_objects_callback(const geometry_msgs::msg::PointStamped::SharedPtr /* msg */)
  {
    if (state_ == STATE_SEARCHING) {
      RCLCPP_INFO(this->get_logger(), "🗑️ ゴミを検出しました。タスクキューに追加します。");
      target_queue_.push_back("dummy_target_1");
      set_state(STATE_APPROACHING);
    }
  }

  void chaser_completion_callback(const std_msgs::msg::Bool::SharedPtr msg)
  {
    if (msg->data && state_ == STATE_APPROACHING) {
      RCLCPP_INFO(this->get_logger(), "👍 接近完了通知を受信しました。ホースでの回収を開始します。");
      set_state(STATE_COLLECTING);
    }
  }

  void hose_result_callback(const std_msgs::msg::Bool::SharedPtr msg)
  {
    if (state_ == STATE_COLLECTING) {
      if (msg->data) {
        RCLCPP_INFO(this->get_logger(), "👍 回収成功！");
      } else {
        RCLCPP_INFO(this->get_logger(), "😥 回収失敗...");
      }
      set_state(STATE_STOPPING);
    }
  }
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<TaskManagerNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
